[do not merge]server 4.0 docs release branch (#7023)
circleci/circleci-docs: jekyll/_cci2/server/installation/phase-1-prerequisites.adoc

* link:https://aws.amazon.com/s3/[AWS S3]

* link:https://min.io/[MinIO]

* link:https://cloud.google.com/storage/[Google Cloud Storage]

While any S3 compatible object storage may work, we test and support AWS S3 and MinIO. For object storage providers that do not support the S3 API, such as Azure blob storage, we recommend using MinIO Gateway.

Follow the instructions below to create a bucket and access method for S3 or GCS.

NOTE: If you are installing behind a proxy, object storage should be behind this proxy also. Otherwise, proxy details will need to be supplied at the job level within every project `.circleci/config.yml` to allow artifacts, test results, cache save and restore, and workspaces to work. For more information see the https://circleci.com/docs/2.0/server/operator/installing-server-behind-a-proxy/[Installing Server Behind a Proxy] guide.

ifndef::env-gcp[]

[#s3-storage]
=== S3 Storage

[#create-aws-s3-bucket]
==== Create AWS S3 Bucket

[source,shell]
----
aws s3api create-bucket \
    --bucket <YOUR_BUCKET_NAME> \
    --region <YOUR_REGION> \
----

> NOTE: `us-east-1` does not support a LocationConstraint. If your region is `us-east-1`, omit the bucket configuration

Author: ```suggestion
```

+
[source,shell]
----
gcloud container clusters create circleci-server \
  --num-nodes 3 \
  --machine-type n1-standard-4 \
  --workload-pool=<PROJECT_ID>.svc.id.goog
----
+
NOTE: Your kube-context should get updated with the new cluster credentials automatically.
+
If you need to update your kube-context manually, you can by running the following:
+
[source,shell]
----
gcloud container clusters get-credentials circleci-server
----
. Install the https://cloud.google.com/blog/products/containers-kubernetes/kubectl-auth-changes-in-gke[GKE authentication plugin] for `kubectl`:
+
[source,shell]
----
gcloud components install gke-gcloud-auth-plugin
----
. Verify your cluster:
+
[source,shell]
----
kubectl cluster-info
----
. Create a service account:
+
[source,shell]
----
gcloud iam service-accounts create <SERVICE_ACCOUNT_ID> --description="<DESCRIPTION>" \
  --display-name="<DISPLAY_NAME>"
----
. Retrieve credentials for the service account:
+
[source,shell]
----
gcloud iam service-accounts keys create <KEY_FILE> \
  --iam-account <SERVICE_ACCOUNT_ID>@<PROJECT_ID>.iam.gserviceaccount.com
----
endif::env-aws[]

[#enable-workload-identities-in-gke]

> ==== Enable Workload Identities in GKE (optional)

Author: I would like to move this section to another page, outside the flow of the installation guide if that will work. To avoid confusing people who have just followed the steps to set up their cluster. I'm wondering if we will potentially have customers who want to use an existing cluster without workload identities who will be following these steps and need this section at this point? I propose to make a page for this content in the operators guide and link to it at this point in the installation guide, something like: "If you already have a cluster you wish to use and want to enable workload identities, see the ***** guide." @circleci/server

Server v4.x hosts build artifacts, test results, and other state object storage. The following storage options are supported:

* link:https://aws.amazon.com/s3/[AWS S3]

* link:https://min.io/[MinIO]

* link:https://cloud.google.com/storage/[Google Cloud Storage]

While any S3 compatible object storage may work, we test and support AWS S3 and MinIO. For object storage providers that do not support S3 API, such as Azure blob storage, we recommend using MinIO Gateway.

Follow the instructions below to create a bucket and access method for S3 or GCS.

NOTE: If you are installing behind a proxy, object storage should be behind this proxy also. Otherwise proxy details will need to be supplied at the job level within every project `.circleci/config.yml` to allow artifacts, test results, cache save and restore, and workspaces to work. For more information see the https://circleci.com/docs/2.0/server-3-operator-proxy/[Configuring a Proxy] guide.

ifndef::env-gcp[]

[#configuring-s3-storage]
=== Configuring S3 Storage

[#create-aws-s3-bucket]
==== Create AWS S3 Bucket

[source,shell]
----
aws s3api create-bucket \
    --bucket <YOUR_BUCKET_NAME> \
    --region <YOUR_REGION> \
----

> NOTE: `us-east-1` does not support a LocationConstraint. If your region is `us-east-1`, omit the bucket configuration

Reviewer: `LocationConstraint` is not in the command above any more it seems.  This note can be removed

---
version:
- Server v4.x
- Server Admin
---
= Phase 1 - Prerequisites
:page-layout: classic-docs
:page-liquid:
:page-description: Find the general and infrastructure-specific requirements that are needed in order to configure the CircleCI server application.
:icons: font
:toc: macro
:toc-title:
:relfileprefix: ../
:relfilesuffix: /

// This doc uses ifdef and ifndef directives to display or hide content specific to Google Cloud Storage (env-gcp) and AWS (env-aws). Currently, this affects only the generated PDFs. To ensure compatability with the Jekyll version, the directives test for logical opposites. For example, if the attribute is NOT env-aws, display this content. For more information, see https://docs.asciidoctor.org/asciidoc/latest/directives/ifdef-ifndef/.

> CircleCI server v4.x is installed as a helm chart. The installation process is broken don into four phases. There is a validation step at the end of each phase, allowing you to confirm success before moving to the next phase. Depending on your requirements, phases 3 and 4 may include multiple steps. This installation guide assumes you have already read the  https://circleci.com/docs/2.0/server/overview/overview/[CircleCI Server v4.x Overview].

Reviewer 2: "CircleCI Server v4.x Overview" is going to a 404 as well.

```suggestion
CircleCI server v4.x is installed as a helm chart. The installation process is broken down into four phases. There is a validation step at the end of each phase, allowing you to confirm success before moving to the next phase. Depending on your requirements, phases three and four may include multiple steps. This installation guide assumes you have already read the https://circleci.com/docs/2.0/server/overview/overview/[CircleCI Server v4.x Overview].
```

.Installation Experience Flow Chart Phase 1
image::server-install-flow-chart-phase1.png[Flow chart showing the installation flow for server 3.x with phase 1 highlighted]

> NOTE: In the following sections, replace any items or credentials displayed between `< >` with your details.

Reviewer 2: Same comment as before, "replace any items or credentials displayed between `< >` with your details" makes it seem the carrots should stay. Maybe rephrase to "replace any items or credentials indicated by `< >` with your details"

[#install-required-software]
== 1. Install required software
Download and install the following software before continuing:

[.table.table-striped]
[cols=4*, options="header", stripes=even]
|===
| Tool
| Version
| Used for
| Notes

| https://www.terraform.io/downloads.html[Terraform]

> | {terraformversion} or greater

Reviewer 2: Should all of these instances in this table between brackets, like {terraformversion}, be in backticks?

Author: These are all variables coming from the jekyll config file so render correctly

| Infrastructure Management
|

| https://kubernetes.io/docs/tasks/tools/install-kubectl/[kubectl]
| {kubectlversion} or greater
| Kubernetes CLI
|

| https://helm.sh/[Helm]
| {helmversion} or greater
| Kubernetes Package Management
|

| https://github.com/databus23/helm-diff[Helm Diff]
| {helmdiffversion} or greater

> | Helping with `values.yaml`` changes and updates

Reviewer 2: It was confirmed by an engineer that we don't actually support .yaml files even though everyone was told we support both .yml and .yaml. Not sure if that matters here.

```suggestion
| Helping with `values.yaml` changes and updates
```

[#create-a-kubernetes-cluster]
== 2. Create a Kubernetes cluster
CircleCI server installs into an existing Kubernetes cluster. The application uses a large number of resources. Depending on your usage, your Kubernetes cluster should meet the following requirements:

[.table.table-striped]
[cols=5*, options="header", stripes=even]
|===
| Number of daily active CircleCI users
| Minimum Nodes
| Total CPU
| Total RAM
| NIC speed

| < 500
| 3
| 12 cores
| 32 GB
| 1 Gbps

| 500+
| 3
| 48 cores
| 240 GB
| 10 Gbps
|===

**Supported Kubernetes versions:**

[.table.table-striped]
[cols=2*, options="header", stripes=even]
|===
| CircleCI Version
| Kubernetes Version

| 4.0.0
| 1.16 - 1.23
|===

Creating a Kubernetes cluster is your responsibility. Please note:

* You must have appropriate permissions to list, create, edit, and delete pods in your cluster. Run this command to verify your permissions:
+
[source,shell]
----
kubectl auth can-i <list|create|edit|delete> pods
----
* There are no requirements regarding VPC setup or disk size for your cluster. It is recommended that you set up a new VPC rather than use an existing one.

ifndef::env-gcp[]

[#eks]
=== EKS

> You can learn more about creating an Amazon EKS cluster https://aws.amazon.com/quickstart/architecture/amazon-eks/[here]. We recommend using https://docs.aws.amazon.com/eks/latest/userguide/getting-started-eksctl.html[`eksctl`] to create your cluster, which creates a VPC and selects the proper security groups for you.

Reviewer 2: Maybe not use the word "here" for a link?

. https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html[Install] and https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html[configure] the AWS CLI for your AWS account.
. Install https://docs.aws.amazon.com/eks/latest/userguide/eksctl.html[`eksctl`].
. Create your cluster by running the following (Cloud formation with `eksctl` and EKS can take more than 20 minutes to complete):
+
[source,shell]
----
eksctl create cluster --name=circleci-server --nodes 4 --node-type m5.xlarge
----
. Once the cluster has been created, you can use the following command to configure `kubectl` access:
+
[source,shell]
----
eksctl utils write-kubeconfig --cluster circleci-server
----

NOTE: You may see the following error `AWS STS Access - cannot get role ARN for current session: InvalidClientTokenID`. This means your AWS credentials are invalid, or your IAM user does not have permission to create an EKS cluster. Proper IAM permissions are necessary to use `eksctl`. See the AWS documentation on https://aws.amazon.com/iam/features/manage-permissions/[IAM permissions].

endif::env-gcp[]

ifndef::env-aws[]

[#gke]
=== GKE

> You can learn more about creating a GKE cluster https://cloud.google.com/kubernetes-engine/docs/how-to#creating-clusters[here].

Reviewer 2: Maybe not use the word "here" for a link?

----
. Verify your cluster:
+
[source,shell]
----
kubectl cluster-info
----
. Create a service account:
+
[source,shell]
----
gcloud iam service-accounts create <SERVICE_ACCOUNT_ID> --description="<DESCRIPTION>" \
  --display-name="<DISPLAY_NAME>"
----
. Retrieve credentials for the service account:
+
[source,shell]
----
gcloud iam service-accounts keys create <KEY_FILE> \
  --iam-account <SERVICE_ACCOUNT_ID>@<PROJECT_ID>.iam.gserviceaccount.com
----
endif::env-aws[]

[#enable-workload-identity-in-gke]
==== Enable Workload Identity in GKE (optional)
Follow these steps if you already have a GKE cluster and need to enable Workload Identity on the cluster and the node pools.

. Enable Workload Identity on existing cluster:
+
[source,shell]
----
  gcloud container clusters update "<CLUSTER_NAME>" \
    --workload-pool="<PROJECT_ID>.svc.id.goog"
----
. Get node pools of existing GKE cluster:
+
[source,shell]
----
  gcloud container node-pools list --cluster "<CLUSTER_NAME>"
----

. Update existing node pools:
+
[source,shell]
----
  gcloud container node-pools update "<NODEPOOL_NAME>" \
    --cluster="<CLUSTER_NAME>" \
    --workload-metadata="GKE_METADATA"
----

> You must repeat Step 3 for all the existing node pools. Follow these links for steps to enable Workload Identity for your Kubernetes service accounts:

Reviewer 2: All the links in this next section go to a 404

* link:https://circleci.com/docs/2.0/server/installation/phase-3-execution-environments/#nomad-autoscaler-gcp[Nomad Autoscaler]
* link:https://circleci.com/docs/2.0/server/installation/phase-3-execution-environments/#gcp-3[VM]
* link:https://circleci.com/docs/2.0/server/installation/phase-1-prerequisites/#configuring-google-cloud-storage[Object-Storage]

[#create-a-new-github-oauth-app]
== 3. Create a new GitHub OAuth app

CAUTION: If GitHub Enterprise and CircleCI server are not on the same domain, then images and icons from GHE will fail to load in the CircleCI web app.

> Registering and setting up a new GitHub OAuth app for CircleCI server allows for authorization control to your server installation using GitHub OAuth and for updates to GitHub projects/repos using build status information. THe following steps apply for both GitHub.com and GitHub Enterprise.

Reviewer 2: ```suggestion
Registering and setting up a new GitHub OAuth app for CircleCI server allows for authorization control to your server installation using GitHub OAuth and for updates to GitHub projects/repos using build status information. The following steps apply for both GitHub.com and GitHub Enterprise.
```

CAUTION: Store these values securely. If they are lost, job history and artifacts will not be recoverable.

[#artifact-signing-key]
=== a. Artifact signing key
To generate an artifact signing key, run the following command:

[source,shell]
----
docker run circleci/server-keysets:latest generate signing -a stdout
----

[#encryption-signing-key]
=== b. Encryption signing key
To generate an encryption signing key, run the following command:

[source,shell]
----
docker run circleci/server-keysets:latest generate encryption -a stdout
----

[#object-storage-and-permissions]
== 6. Object storage and permissions
Server v4.x hosts build artifacts, test results, and other state object storage. The following storage options are supported:

* link:https://aws.amazon.com/s3/[AWS S3]

* link:https://min.io/[MinIO]

* link:https://cloud.google.com/storage/[Google Cloud Storage]

While any S3 compatible object storage may work, we test and support AWS S3 and MinIO. For object storage providers that do not support the S3 API, such as Azure blob storage, we recommend using MinIO Gateway.

Follow the instructions below to create a bucket and access method for S3 or GCS.

> NOTE: If you are installing behind a proxy, object storage should be behind this proxy also. Otherwise, proxy details will need to be supplied at the job level within every project `.circleci/config.yml` to allow artifacts, test results, cache save and restore, and workspaces to work. For more information see the https://circleci.com/docs/2.0/server/installation/installing-server-behind-a-proxy/[Installing Server Behind a Proxy] guide.

Reviewer 2: The Installing Server Behind a Proxy link in this section goes to a 404.

----

. Get the OIDC provider ARN. You will need this in later steps.
+
[source,shell]
----
aws iam list-open-id-connect-providers | grep $(aws eks describe-cluster --name <CLUSTER_NAME> --query "cluster.identity.oidc.issuer" --output text | awk -F'/' '{print $NF}')
----

. Get your OIDC provider URL. You will need this in later steps.
+
[source,shell]
----
aws eks describe-cluster --name <CLUSTER_NAME> --query "cluster.identity.oidc.issuer" --output text | sed -e "s/^https:\/\///"
----

. Create the role using the command and trust policy template below. You will need the Role ARN and name in later steps.
+
[source,shell]
----
aws iam create-role --role-name circleci-s3 --assume-role-policy-document file://<TRUST_POLICY_FILE>
----
+
[source, json]
----
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Federated": "<OIDC_PROVIDER_ARN>"
      },
      "Action": "sts:AssumeRoleWithWebIdentity",
      "Condition": {
        "StringEquals": {
          "<OIDC_PROVIDER_URL>:sub": "system:serviceaccount:<K8S_NAMESPACE>:object-storage"
        }
      }
    }
  ]
}
----

> . Create the policy using the command and template below.  Fill in the bucket name and the role ARN.

Reviewer 2: ```suggestion
. Create the policy using the command and template below. Fill in the bucket name and the role ARN.
```

        "s3:GetBucketVersioning",
        "s3:GetBucketAcl",
        "s3:PutInventoryConfiguration",
        "s3:GetObjectTorrent",
        "s3:PutBucketWebsite",
        "s3:PutBucketRequestPayment",
        "s3:PutObjectRetention",
        "s3:GetBucketCORS",
        "s3:GetBucketLocation",
        "s3:GetAccessPointPolicy",
        "s3:GetObjectVersion",
        "s3:GetAccessPoint",
        "s3:GetAccountPublicAccessBlock",
        "s3:ListAllMyBuckets",
        "s3:ListAccessPoints",
        "s3:ListJobs"
      ],
      "Resource": [
        "arn:aws:s3:::<YOUR_BUCKET_NAME>",
        "arn:aws:s3:::<YOUR_BUCKET_NAME>/*"
      ]
    },
    {
      "Effect": "Allow",
      "Action": [
        "iam:GetRole",
        "sts:AssumeRole"
      ],
      "Resource": "<OBJECT_STORAGE_ROLE_ARN>"
    }
  ]
}
----

. Attach the policy to the role:
+
[source,shell]
----
aws iam attach-role-policy --role-name <OBJECT_STORAGE_ROLE_NAME> --policy-arn=<STORAGE_POLICY_ARN>
----
--

[.tab.authentication.IAM_access_keys]
**Option 2:** IAM access keys

. Create an IAM user for CircleCI server.
+
[source,shell]
----
aws iam create-user --user-name circleci-server
----

> . Create a policy document _policy.json_.

Reviewer 2: ```suggestion
. Create a policy document `policy.json`.
```

----
aws iam put-user-policy \
  --user-name circleci-server \
  --policy-name circleci-server \
  --policy-document file://policy.json
----

. Create Access Key for user circleci-server.
+
NOTE: You will need this when you configure your server installation later.
+
[source,shell]
----
aws iam create-access-key --user-name circleci-server
----
+
The result should look like this:
+
[source, json]
----
{
  "AccessKey": {
        "UserName": "circleci-server",
        "Status": "Active",
        "CreateDate": "2017-07-31T22:24:41.576Z",
        "SecretAccessKey": <AWS_SECRET_ACCESS_KEY>,
        "AccessKeyId": <AWS_ACCESS_KEY_ID>
  }
}
----

endif::env-gcp[]

ifndef::env-aws[]

[#google-cloud-storage]
=== Google Cloud Storage

[#create-a-gcp-bucket]
==== a. Create a GCP bucket
If your server installation runs in a GKE cluster, ensure that your current IAM user is cluster admin for this cluster, as RBAC (role-based access control) objects need to be created. More information can be found in the https://cloud.google.com/kubernetes-engine/docs/how-to/role-based-access-control[GKE documentation].

[source,shell]
----
gsutil mb gs://circleci-server-bucket
----

[#set-up-authentication]

> ==== b. Set up authentication

Reviewer 2: This section has the same name "Set up authentication" which causes the ToC nav to show the user is on both of them (they are both highlighted blue). Not an issue for here, but probably something we should bring up with Docs Disco to see if we can fix that. It's reasonable to think that some sections will have the exact same names, particularly when a user needs to follow the same steps for different services.
